sling/pyapi/pyframe.cc

```diff
@@ -0,0 +1,365 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "sling/pyapi/pyframe.h"
+
+#include "sling/pyapi/pystore.h"
+
+namespace sling {
+
+// Python type declarations.
+PyTypeObject PyFrame::type;
+PyMappingMethods PyFrame::mapping;
+PySequenceMethods PyFrame::sequence;
+PyTypeObject PySlots::type;
+
+PyMethodDef PyFrame::methods[] = {
+  {"data", (PyCFunction) &PyFrame::Data, METH_KEYWORDS, ""},
+  {"append", (PyCFunction) &PyFrame::Append, METH_VARARGS, ""},
+  {"extend", (PyCFunction) &PyFrame::Extend, METH_O, ""},
+  {"store", (PyCFunction) &PyFrame::GetStore, METH_NOARGS, ""},
+  {nullptr}
+};
+
+void PyFrame::Define(PyObject *module) {
+  InitType(&type, "sling.Frame", sizeof(PyFrame));
+  type.tp_dealloc = reinterpret_cast<destructor>(&PyFrame::Dealloc);
+  type.tp_getattro = &PyFrame::GetAttr;
+  type.tp_setattro = &PyFrame::SetAttr;
+  type.tp_str = &PyFrame::Str;
+  type.tp_iter = &PyFrame::Slots;
+  type.tp_call = &PyFrame::Find;
+  type.tp_hash = &PyFrame::Hash;
+  type.tp_richcompare = &PyFrame::Compare;
+  type.tp_methods = methods;
+
+  type.tp_as_mapping = &mapping;
+  mapping.mp_length = &PyFrame::Size;
+  mapping.mp_subscript = &PyFrame::Lookup;
+  mapping.mp_ass_subscript = &PyFrame::Assign;
+
+  type.tp_as_sequence = &sequence;
+  sequence.sq_contains = &PyFrame::Contains;
+
+  RegisterType(&type);
+}
+
+void PyFrame::Init(PyStore *pystore, Handle handle) {
+  // Add frame as root object for store to keep it alive in the store.
+  InitRoot(pystore->store, handle);
+
+  // Add reference to store to keep it alive.
+  this->pystore = pystore;
+  Py_INCREF(pystore);
+}
+
+void PyFrame::Dealloc() {
+  // Unlock tracking of handle in store.
+  Unlink();
+
+  // Release reference to store.
+  Py_DECREF(pystore);
+}
+
+Py_ssize_t PyFrame::Size() {
+  return frame()->slots();
+}
+
+long PyFrame::Hash() {
+  return handle().bits;
+}
+
+PyObject *PyFrame::Compare(PyObject *other, int op) {
+  // Only equality check is suported.
+  if (op != Py_EQ && op != Py_NE) {
+    PyErr_SetString(PyExc_TypeError, "Invalid frame comparison");
+    return nullptr;
+  }
+
+  // Check if other object is a frame.
+  bool match = false;
+  if (PyObject_TypeCheck(other, &PyFrame::type)) {
+    // Check if the stores and handles are the same.
+    PyFrame *pyother = reinterpret_cast<PyFrame *>(other);
+    match = pyother->pystore->store == pystore->store &&
+            pyother->handle() == handle();
+  }
+
+  if (op == Py_NE) match = !match;
+  if (match) {
+    Py_RETURN_TRUE;
+  } else {
+    Py_RETURN_FALSE;
+  }
+}
+
+PyObject *PyFrame::GetStore() {
+  Py_INCREF(pystore);
```

Comment:
Is this refcount decremented when the python object that stores this reference expires?

Reply:
Yes. In general, when you return a Python object, the caller will have a new reference to it.